update-chart-versions.sh

```diff
@@ -0,0 +1,168 @@
+pathToChart="./$chartDirectory/Chart.yaml"
+
+namePrefix="- name: "
+versionPrefix="  version: "
+
+localComponentNames=()
+localComponentVersions=()
+
+versionSource='--devel'
+
+
+# extract options and their arguments into variables.
+while true ; do
+  case "$1" in
+    -h | --help )
+      displayHelp
+      shift
+      ;;
+    -m | --master )
+      unset versionSource
+      shift
+      ;;
+    -p | --passed-args-only )
+      usePassedArgsOnly="true"
+      shift
+      ;;
+    -- )
+      shift
+      break
+      ;;
+    *)
+      echo "Internal error!"
+      exit 1
+      ;;
+  esac
+done
+
+# Update the local cache from the Repo and confirm dev repo is queried
+echo -e "Updating local repo cache... \n"
+updateRepos=$(helm repo update|grep "gpitfuturesdevacr")
+
+if [[ $updateRepos != "" ]]; then
+  echo -e "$updateRepos \n"
+else
+  echo -e "gpitfuturesdevacr not found in helm repos, please add 'gpitfuturesdevacr' to your helm repos and try again.  \n"
+  exit
+fi
+
+# Construct an indexed array of our components & their versions
+while IFS= read -r line; do
+  if [[ $line =~ ^"$namePrefix"* ]]; then
+      localComponentNames+=($(echo "${line#${namePrefix}}"))
+  fi
+
+  if [[ $line =~ ^"$versionPrefix"* ]]; then
+      localComponentVersions+=($(echo "${line#${versionPrefix}}"))
+  fi
+done < "$pathToChart"
+
+if [ "${#localComponentNames[@]}" -ne "${#localComponentVersions[@]}" ]; then
+  >&2 echo "Number of components doesn't match the number of versions, exiting."
+  exit
+fi
+
+# build a map (chart => version) for local charts, eg:
+# of => 1.27.1
+# mp => 1.27.1
+declare -A localChartsToVersions
+for (( i=0; i<${#localComponentNames[@]}; i++ )); do
+  localChartsToVersions["${localComponentNames[$i]}"]="${localComponentVersions[$i]}"
+done
+
+# build a map (chart => version) for remote charts
+if [ -z $usePassedArgsOnly ]; then
+  declare -A remoteChartsToVersions
+  echo -e "Grabbing Chart information from the acr... \n"
+  remoteChartsAndVersions=$(helm search repo gpit $versionSource | sed "s@gpitfuturesdevacr/@@" | awk 'NR>1{printf("%s:%s ", $1, $2)}')
+  for entry in ${remoteChartsAndVersions[*]}; do
+    component=$(echo $entry | cut -d: -f1 ) 
+    version=$(echo $entry | cut -d: -f2 ) 
+    remoteChartsToVersions[$component]=$version
+  done
+fi
+
+# create a copy of the localChartsToVersions map that will serve as a change set
+declare -A changeSet  
+for key in "${!localChartsToVersions[@]}"; do
+  changeSet["$key"]="${localChartsToVersions["$key"]}"
+done
+
+for component in ${!localChartsToVersions[@]}; do
+  if [ "${remoteChartsToVersions[$component]+isset}" ]; then
+    localVersion="${localChartsToVersions[$component]}"
+    remoteVersion="${remoteChartsToVersions[$component]}"
+    if [ "$localVersion" != "$remoteVersion" ]; then
+        changeSet[$component]=$remoteVersion
+    fi
+  fi
+done
+
+# Parse any extra ags, add the versions that need to be changed to the changeSet
+for argument in "$@"; do
```

Comment:
can't this be combined with the other arg parsing above?

Reply:
It could, we would have to introduce another construct to hold whats in $@ though.
It is going to be a little extra work but I suppose it will improve the readability of the script.